### ISSUE TITLE
Make IPTS compatible with 4.10 kernel

### ISSUE THREAD

**Issue by axelrtgs** [ipts-linux-org/ipts-linux-new#3]
Hello,

It would be nice if this was compatible with 4.10 kernel due to it including some much needed drivers and fixes for Microsoft Surface devices. the only thing i needed to do with the 4.10 kernel to get it to work on the Surface Book is to compile the kernel with wifi power save disabled to make the wifi stable (still testing long term stability)

For my own purposes i have tried merging your IPTS specific commits into the 4.10-rc3 kernel but due to some small changes i had to make some small modifications to get it to compile. I have not tested the changes yet (will do so tonight) but iI' sure i screwed something up since i dont know the first thing about C.

If you could review the inclusion into the 4.10-rc3 kernel and see if i mucked anything up it would be appreciated, you can find [my repo here](https://github.com/axelrtgs/linux-kernel-ipts-4.10) the actual proper kernel is in the branch, master is just the vanilla ubuntu mainline kernel with no modifications.

Feel free to use the commit patch to merge into your own ipt-linux-new repo on the updated kernel it is after all your hard work.

hoping you have time to take a look and fix any bugs i may have introduced,
thank you,
Jamie

**Comment by afzalarsalan** [comment]
Well the coincidence is uncanny because I've also been working on rebasing this patch for linux 4.10 and have more or less ended up with the same changes as you. The only thing different really is with your addition of client_base which wouldn't really work since it has no ties to the original driver and probably never gets defined. [This commit](https://github.com/torvalds/linux/commit/72aa0d899a3ea7e14436ac0b869cf3333277b92e) seems to have changed client_base into vaddr which also seems to compile for me. Unfortunately, I'm still ending up with a black screen after the i915 module gets loaded so it doesn't seem to make any difference. Let me know if your version has any better success but we might just be stuck in the water until someone at Intel who understands the whole abstraction reworks this code.

**Comment by axelrtgs** [comment]
no i am stuck in the same place also, black screen. thanks for the vaddr tip, i made the change in my code.

i am not a programmer by any means so C is way above my head but perhaps there is something missing. i wish there was a way to debug more efficiently whats going on and where it gets stuck.

maybe some firmware is missing

edit: one of the other changes i made vs the vanilla ipts patch was in two functions remove the & in front of &dev_priv in  function intel_lr_context_pin & execlists_context_deffered_alloc of intel_ipts.c 

Edit 2: found a place where we can get 'extra' firmware from the linux team not included in the linux kernel due to it not being gpl but allowing redistribution. i have downloaded it and added the ipts .bin files needed for the kernel ipts driver to work. all you need to do is clone the repo and run 'make check' to see if everything is there and 'sudo make install' to copy the firmware over to the right place including some other ones that may be good to have (like GUC is included in this package also) [https://github.com/axelrtgs/linux-firmware-ipts](https://github.com/axelrtgs/linux-firmware-ipts) is where you can find the extra firmware with the ipts mod. NOTE: i am just comiting it here so i dont lose my work but this is a personal repo not meant for redistribution.

Edit 3: one of the programmers at work familiar with C is helping me in his free time to look into this problem. got a few things confirmed in the modifications i did which indicate they make sense, now im recompiling and then will test tonight. i enabled the debug line in the driver so i can get some additional info and see whats going on after it gets stuck

Edit4: i took a look at your repo, i am comiting the wifi kernel crash fix now. i was experiencing that issue as well so hopefully this fixes it. on 4.10 i dont think the rest of the patches are important, i will check it out once i can get the wifi stable and the crash to stop and see whats not working. hopefully i can figure out the touch thing soon.

**Comment by axelrtgs** [comment]
@afzalarsalan I figured this deserved it own post. ive got it booting on my build and working WITH touch on 4.10rc3. just recompiling with the wifi patch to fix the wifi crash and then see how it goes.

keep you posted if anything comes up.

**Comment by afzalarsalan** [comment]
Let me know what commits you end up using in the end, genuinely curious what the secret sauce is. Also in regards to wifi, you don't need to disable powersaving as from my testing it seems to make a rather minimal amount of difference in ping and no change at all in stability.

**Comment by axelrtgs** [comment]
i disabled the powersaving because before the wifi fix it made wifi more stable and it seems to have been a common theme in the past. feel free to re-enable it. as for stability it looks pretty stable just don't enable the multitouch for ipts the kernal hangs immediately. i have re-created the repo based on a cleaner structure, the previous one due to all the .gitignore files was missing a bunch of stuff so wouldn't compile. now it should be complete. you can check the commits but they are similar to before with the exception of fixing the client_base as you thankfully pointed out.

the secret sauce i think is in the firmware package which may have resolved some missing dependency. try it out on your kernel and see if it works.

EDIT: tried recompiling the fresh branch and it didnt work, only difference is in the misc/ipts/ drivers i enable the debug mode (uncomment the define) recompiling and trying it now again. this might be the secret sauce.

**Comment by axelrtgs** [comment]
Confirmed enabling the debug define makes the driver work. not sure why but it does. ill investigate tomorrow with my colleague from work who knows C.

only thing i notice is i get the following messages from the kernel and not sure if they are important or not:


[    6.053634] i2c_hid i2c-MSHW0030:00: error in i2c_hid_init_report size:63 / ret_size:0
[    6.055727] i2c_hid i2c-MSHW0030:00: error in i2c_hid_init_report size:63 / ret_size:0
[    6.056630] i2c_hid i2c-MSHW0030:00: error in i2c_hid_init_report size:19 / ret_size:0
[    6.057523] i2c_hid i2c-MSHW0030:00: error in i2c_hid_init_report size:19 / ret_size:63
[    6.058346] i2c_hid i2c-MSHW0030:00: error in i2c_hid_init_report size:19 / ret_size:7
[    6.059337] i2c_hid i2c-MSHW0030:00: error in i2c_hid_init_report size:19 / ret_size:12
[    6.060159] i2c_hid i2c-MSHW0030:00: error in i2c_hid_init_report size:19 / ret_size:0
[    6.061003] i2c_hid i2c-MSHW0030:00: error in i2c_hid_init_report size:19 / ret_size:0
[    6.061865] i2c_hid i2c-MSHW0030:00: error in i2c_hid_init_report size:19 / ret_size:4
[    6.062734] i2c_hid i2c-MSHW0030:00: error in i2c_hid_init_report size:19 / ret_size:15

**Comment by afzalarsalan** [comment]
That's absolutely hilarious, wonder how that works. Also those errors are related to some other sensors and from what I can tell are harmless. Haven't found the time to track down the issue with those but it's irrelevant to this.

**Comment by axelrtgs** [comment]
My wifi is still unstable. The wifi fix, fixed the kernel crash but I still lose all connectivity even though the manager shows connected and the Smersh log just keeps repeating that the wifi interface is already connected. 

I remember reading about that somewhere I'll try and dig it up. 

If you have any ideas please share.

**Comment by afzalarsalan** [comment]
Yeah this used to be a problem back in 4.8 funny enough, the kernel didn't crash back then but connectivity would go in and out. Linux 4.10 definitely shouldn't have regressed on that issue especially since our driver development is a bit on the slower side compared to others. Will test on my end after kernel builds with touch to see if this issue is back. Try checking dmesg and see if you see anything out of the norm when you lose connectivity

**Comment by afzalarsalan** [comment]
Also does touch actually work for you, I have it booting up with the patch changes you made but I don't end up with actual touch like in 4.9.

**Comment by axelrtgs** [comment]
Yes the touch is working. On my surface book I am using the 79.bin flavor of the firmware. If you checkout my Linux-firmware package and install it should resolve the issue. Maybe your missing a dependency, I know I probably was but the package fixed it

**Comment by afzalarsalan** [comment]
Did you ever figure out why debug is required in order to make IPTS work with booting? Still find that quite peculiar

**Comment by Madwonk** [comment]
@axelrtgs : I left a mention of your 4.10 kernel on the [Ubuntu bug report](https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1652815). It'd be nice if we can get some attention directed towards these bug reports so this could get merged in.

**Comment by afzalarsalan** [comment]
While not being the necessarily the person you were addressing, I'm pretty sure we would both agree that this code in no where near merge-ready if simply for the fact that the original reasons provided by Intel for lack of merger have failed to be addressed. Primarily because the conditions to make this this merge ready would require more work on not only this driver but the i915 display driver than honestly anyone should be willing to do for free time. It's not exactly something you can pump out in a couple nights. The implementation on the 4.10 already is sketchy as is and I have yet to even try to merge this into the 4.11 tree much less the soon to be 4.12 head tree.

**Comment by axelrtgs** [comment]
yeah i agree, im am in no way a devloper and my hacks in the code to get it to work may cause many adverse effect. i would not consider this merge to kernel ready but maybe just aguide to how i got it working and point someone in the right direction

**Comment by Madwonk** [comment]
@afzalarsalan True, a very good point. There is no way this code is ready in its current state. The hope I have is, however, that more discussion about it will lead to more attention by the people who actually know what they're doing.

**Comment by oleid** [comment]
For people, which are not aware of it: Here is a repo with ipts for 4.14:

https://github.com/jakeday/linux-surface

**Comment by thekswenson** [comment]
Thanks for the tip @oleid !

**Comment by skymage23** [comment]
Thanks. I'll try it as soon as I can.
